include/exadg/incompressible_navier_stokes/postprocessor/postprocessor.cpp

```diff
@@ -182,105 +191,123 @@ void
-    this->additional_fields.push_back(sol);
+    streamfunction.type        = SolutionFieldType::scalar;
+    streamfunction.name        = "streamfunction";
+    streamfunction.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
+    navier_stokes_operator->initialize_vector_velocity_scalar(
+      streamfunction.get_vector_reference());
+    streamfunction.recompute_solution_field = [&](VectorType & dst, VectorType const &) {
+      navier_stokes_operator->compute_streamfunction(dst, evaluate_get(vorticity));
+    };
+
+    this->additional_fields_vtu.push_back(&streamfunction);
   }
 
   // q criterion
-  if(pp_data.output_data.write_q_criterion == true)
+  if(pp_data.output_data.write_q_criterion)
   {
-    navier_stokes_operator->initialize_vector_velocity_scalar(q_criterion);
-
-    SolutionField<dim, Number> sol;
-    sol.type        = SolutionFieldType::scalar;
-    sol.name        = "q_criterion";
-    sol.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
-    sol.vector      = &q_criterion;
-    this->additional_fields.push_back(sol);
+    q_criterion.type        = SolutionFieldType::scalar;
+    q_criterion.name        = "q_criterion";
+    q_criterion.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
+    navier_stokes_operator->initialize_vector_velocity_scalar(q_criterion.get_vector_reference());
+    q_criterion.recompute_solution_field = [&](VectorType & dst, VectorType const & src) {
+      navier_stokes_operator->compute_q_criterion(dst, src);
+    };
+
+    this->additional_fields_vtu.push_back(&q_criterion);
   }
 
   // mean velocity
-  if(pp_data.output_data.mean_velocity.is_active == true)
+  if(pp_data.output_data.mean_velocity.is_active)
   {
-    navier_stokes_operator->initialize_vector_velocity(mean_velocity);
-
-    SolutionField<dim, Number> sol;
-    sol.type        = SolutionFieldType::vector;
-    sol.name        = "mean_velocity";
-    sol.dof_handler = &navier_stokes_operator->get_dof_handler_u();
-    sol.vector      = &mean_velocity;
-    this->additional_fields.push_back(sol);
+    mean_velocity.type        = SolutionFieldType::vector;
+    mean_velocity.name        = "mean_velocity";
+    mean_velocity.dof_handler = &navier_stokes_operator->get_dof_handler_u();
+    navier_stokes_operator->initialize_vector_velocity(mean_velocity.get_vector_reference());
+    mean_velocity.recompute_solution_field = [&](VectorType & dst, VectorType const & src) {
+      compute_mean_velocity(dst, src);
+    };
+
+    this->additional_fields_vtu.push_back(&mean_velocity);
   }
 
   // cfl
   if(pp_data.output_data.write_cfl)
   {
-    SolutionField<dim, Number> sol;
-    sol.type   = SolutionFieldType::cellwise;
-    sol.name   = "cfl_relative";
-    sol.vector = &cfl_vector;
-    this->additional_fields.push_back(sol);
+    // TODO: IS THIS A BUG OR DO CELL VECTORS NEED NOT INTIALIZATION?
+    cfl_vector.type                     = SolutionFieldType::cellwise;
```

Comment:
There is no initialization of the Vector. Do Cell vectors need no intialization?